# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1694196593 -7200
# Parent  42ba40e5ea5998a8dbad6e8abad114a349d8ff07
Bug 1845155 - Strip localized Re: prefixes in NNTP messages. r=mkmelin

Some improvements to original Betterbird code by Hartmut Welpmann <h.w.forms@arcor.de>

diff --git a/mailnews/news/src/NntpNewsGroup.jsm b/mailnews/news/src/NntpNewsGroup.jsm
--- a/mailnews/news/src/NntpNewsGroup.jsm
+++ b/mailnews/news/src/NntpNewsGroup.jsm
@@ -126,16 +126,47 @@ class NntpNewsGroup {
         }
       }
       return [start, end];
     }
     return [];
   }
 
   /**
+   * Strip multiple localized Re: prefixes and set the subject and the hasRe
+   * flag. This emulates NS_MsgStripRE()
+   *
+   * @param {nsIMsgDBHdr} msgHdr - The nsIMsgDBHdr to update
+   * @param {string} subject - The unprocessed subject
+   */
+  setSubject(msgHdr, subject) {
+    // Some code from `setSubject()` in enigmailMsgHdrViewOverlay.js.
+    let rawPrefixes = Services.prefs.getComplexValue(
+      "mailnews.localizedRe",
+      Ci.nsIPrefLocalizedString
+    ).data;
+    let prefixes = [];
+    if (rawPrefixes.length > 0) {
+      prefixes = rawPrefixes.split(",");
+    }
+    if (!prefixes.includes("Re")) {
+      prefixes.push("Re");
+    }
+    // Construct a regular expression like this: ^(Re: |Aw: )+
+    let newSubject = subject.replace(
+      new RegExp(`^(${prefixes.join(": |")}: )+`, "i"),
+      ""
+    );
+    msgHdr.subject = newSubject;
+    if (newSubject != subject) {
+      msgHdr.orFlags(Ci.nsMsgMessageFlags.HasRe);
+    }
+  }
+
+  /**
    * Parse an XOVER line to a msg hdr.
    *
    * @param {string} line - An XOVER response line.
    */
   processXOverLine(line) {
     let parts = line.split("\t");
     if (parts.length < 8) {
       return;
@@ -147,17 +178,17 @@ class NntpNewsGroup {
       date,
       messageId,
       references,
       bytes,
       lines,
     ] = parts;
     let msgHdr = this._db.createNewHdr(articleNumber);
     msgHdr.orFlags(Ci.nsMsgMessageFlags.New);
-    msgHdr.subject = subject;
+    this.setSubject(msgHdr, subject);
     msgHdr.author = from;
     msgHdr.date = new Date(date).valueOf() * 1000;
     msgHdr.messageId = messageId;
     msgHdr.setReferences(references);
     msgHdr.messageSize = bytes;
     msgHdr.lineCount = lines;
     this._msgHdrs.push(msgHdr);
   }
@@ -230,17 +261,17 @@ class NntpNewsGroup {
     switch (name) {
       case "from":
         this._msgHdr.author = value;
         break;
       case "date":
         this._msgHdr.date = new Date(value).valueOf() * 1000;
         break;
       case "subject":
-        this._msgHdr.subject = value;
+        this.setSubject(this._msgHdr, value);
         this._msgHdr.orFlags(Ci.nsMsgMessageFlags.New);
         break;
       case "message-id":
         this._msgHdr.messageId = value;
         break;
       case "references":
         this._msgHdr.setReferences(value);
         break;
